### ISSUE TITLE
f16 k/v cache

### ISSUE THREAD

**Issue by flaneur2020** [crabml/crabml#61]
kv cache contributed a big amount of memory consumption on long chat context. before implementing the SOTA solutions like paged attention, we can put the memory foot print much fewer by using f16 instead of f32. 

f16 and f32 are both expected to be supported, but f16 is better to be set as the default.

**Comment by Xuanwo** [comment]
I'm interested in this issue but don't know how to approach it. Could you provide me with some resources or references to get started?

**Comment by flaneur2020** [comment]
sure!

the kv cache is initialized in llama2.rs Llama2Runner:

![Screenshot 2024-03-03 at 11 54 56](https://github.com/crabml/crabml/assets/129800/b31a3e9c-9eef-4314-aac4-7ddac285267b)

we need add an additional `dtype` argument to the `CpuTensor::alloc` method.

the only arithmetic related with kv cache is the `batch_matmul_vec` operator. we need add a specialized `batch_matmul_vec` about multiplying a f16 matrix and a f32 vector (of course we could transform the f32 vector into a f16 vector before doing the multiplication, this may help improve the performance)

so in short i think it can be seperated to 4 parts:

1. add f16 buffer support
2. add a `dtype` parameter to `Tensor::alloc`
3. add specialized batch_matmul_vec on f16 matrix and f32 vector
4. add a cli parameter allow us choose f16 or f32 kv cache (can be sperated in a stand-alone pull request)

**Comment by flaneur2020** [comment]
sorry, I would like to take on this issue to complete my optimization work. as the benchmark shown under https://github.com/crabml/crabml/pull/85 indicates, the f16 kv cache is the main performance gap compared to llama.cpp
